pkg/operator/staticpod/controller/guard/guard_controller.go

```diff
@@ -0,0 +1,254 @@
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/informers"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	policyclientv1 "k8s.io/client-go/kubernetes/typed/policy/v1"
+	corelisterv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/staticpod/controller/guard/bindata"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+// GuardController is a controller that watches amount of static pods on master nodes and
+// renders guard pods with a pdb to keep maxUnavailable to be at most 1
+type GuardController struct {
+	targetNamespace, podResourcePrefix string
+	operatorName                       string
+	readyzPort                         string
+
+	nodeLister corelisterv1.NodeLister
+	podLister  corelisterv1.PodLister
+	podGetter  corev1client.PodsGetter
+	pdbGetter  policyclientv1.PodDisruptionBudgetsGetter
+
+	// installerPodImageFn returns the image name for the installer pod
+	installerPodImageFn   func() string
+	createConditionalFunc resourceapply.ConditionalFunction
+	deleteConditionalFunc resourceapply.ConditionalFunction
+}
+
+func NewGuardController(
+	targetNamespace, podResourcePrefix string,
+	operatorName string,
+	readyzPort string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	kubeInformersClusterScoped informers.SharedInformerFactory,
+	operatorClient operatorv1helpers.StaticPodOperatorClient,
+	podGetter corev1client.PodsGetter,
+	pdbGetter policyclientv1.PodDisruptionBudgetsGetter,
+	eventRecorder events.Recorder,
+	createConditionalFunc resourceapply.ConditionalFunction,
+	deleteConditionalFunc resourceapply.ConditionalFunction,
+) factory.Controller {
+	c := &GuardController{
+		targetNamespace:       targetNamespace,
+		podResourcePrefix:     podResourcePrefix,
+		operatorName:          operatorName,
+		readyzPort:            readyzPort,
+		nodeLister:            kubeInformersClusterScoped.Core().V1().Nodes().Lister(),
+		podLister:             kubeInformersForTargetNamespace.Core().V1().Pods().Lister(),
+		podGetter:             podGetter,
+		pdbGetter:             pdbGetter,
+		installerPodImageFn:   getInstallerPodImageFromEnv,
+		createConditionalFunc: createConditionalFunc,
+		deleteConditionalFunc: deleteConditionalFunc,
```

Comment:
Looking at https://github.com/openshift/cluster-kube-scheduler-operator/pull/373 I see this is only used for SNO check, any other particular reason you want to split into create and delete condition? If that's only SNO, why not having a single check?

Reply:
Since create is not always of !delete (not delete)

Reply:
Right, but returning `bool, errror` tuple is an option you also mention there.